File: src/main/java/com/werfad/finder/Char2Finder.java
package com.werfad.finder;

import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.util.TextRange;
import com.werfad.KeyTagsGenerator;
import com.werfad.MarksCanvas;
import com.werfad.UserConfig;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Char2Finder implements Finder {
    private static final int STATE_WAIT_SEARCH_CHAR1 = 0;
    private static final int STATE_WAIT_SEARCH_CHAR2 = 1;
    private static final int STATE_WAIT_KEY = 2;
    private int state = STATE_WAIT_SEARCH_CHAR1;
    private UserConfig.DataBean config = UserConfig.getDataBean();

    private String s;
    private TextRange visibleRange;

    private char firstChar = ' ';

    @Nullable
    @Override
    public List<MarksCanvas.Mark> start(@NotNull Editor e, @NotNull String s, @NotNull TextRange visibleRange) {
        this.s = s;
        this.visibleRange = visibleRange;
        state = STATE_WAIT_SEARCH_CHAR1;
        return null;
    }

    @Nullable
    @Override
    public List<MarksCanvas.Mark> input(@NotNull Editor e, char c, @NotNull List<MarksCanvas.Mark> lastMarks) {
        switch (state) {
            case STATE_WAIT_SEARCH_CHAR1:
                firstChar = c;
                state = STATE_WAIT_SEARCH_CHAR2;
                return null;
            case STATE_WAIT_SEARCH_CHAR2:
                boolean ignoreCase = config.isSmartcase() && Character.isLowerCase(firstChar) && Character.isLowerCase(c);
                Pattern pattern = Pattern.compile((ignoreCase ? "(?i)" : "") + firstChar + c);
                Matcher m = pattern.matcher(s);

                List<Integer> offsets = new ArrayList<>();
                while (m.find()) {
                    offsets.add(m.start() + visibleRange.getStartOffset());
                }

                offsets.sort((o1, o2) -> {
                    int caretOffset = e.getCaretModel().getOffset();
                    return Math.abs(o1 - caretOffset) - Math.abs(o2 - caretOffset);
                });

                List<String> tags = KeyTagsGenerator.createTagsTree(offsets.size(), config.getCharacters());
                state = STATE_WAIT_KEY;

                List<MarksCanvas.Mark> res = new ArrayList<>();
                for (int i = 0; i < offsets.size(); i++) {
                    res.add(new MarksCanvas.Mark(tags.get(i), offsets.get(i)));
                }
                return res;
            case STATE_WAIT_KEY:
                return FinderHelper.matchInputAndCreateMarks(c, lastMarks);
            default:
                throw new RuntimeException("Impossible.");
        }
    }
}

File: src/main/kotlin/com/werfad/finder/Char2Finder.kt
package com.werfad.finder

import com.intellij.openapi.editor.Editor
import com.intellij.openapi.util.TextRange
import com.werfad.KeyTagsGenerator
import com.werfad.MarksCanvas
import com.werfad.UserConfig
import com.werfad.UserConfig.DataBean
import com.werfad.utils.findAll
import kotlin.math.abs

private const val STATE_WAIT_SEARCH_CHAR1 = 0
private const val STATE_WAIT_SEARCH_CHAR2 = 1
private const val STATE_WAIT_KEY = 2

class Char2Finder : Finder {
    private var state = STATE_WAIT_SEARCH_CHAR1
    private val config: DataBean = UserConfig.getDataBean()
    private lateinit var s: String
    private lateinit var visibleRange: TextRange
    private var firstChar = ' '

    override fun start(e: Editor, s: String, visibleRange: TextRange): List<MarksCanvas.Mark>? {
        this.s = s
        this.visibleRange = visibleRange
        state = STATE_WAIT_SEARCH_CHAR1
        return null
    }

    override fun input(e: Editor, c: Char, lastMarks: List<MarksCanvas.Mark>): List<MarksCanvas.Mark>? {
        return when (state) {
            STATE_WAIT_SEARCH_CHAR1 -> {
                firstChar = c
                state = STATE_WAIT_SEARCH_CHAR2
                null
            }
            STATE_WAIT_SEARCH_CHAR2 -> {
                val caretOffset = e.caretModel.offset
                val offsets = s.findAll("" + firstChar + c, true)
                    .map { it + visibleRange.startOffset }
                    .sortedBy { abs(it - caretOffset) }
                    .toList()

                val tags = KeyTagsGenerator.createTagsTree(offsets.size, config.characters)
                state = STATE_WAIT_KEY
                offsets.zip(tags)
                    .map { MarksCanvas.Mark(it.second, it.first) }
                    .toList()
            }
            STATE_WAIT_KEY -> advanceMarks(c, lastMarks)
            else -> throw RuntimeException("Impossible.")
        }
    }
}